tests/system/apmserver.py

```diff
@@ -91,63 +124,179 @@ def config(self):
+        return super(ClientSideBaseTest, self).get_transaction_payload_path(name)
+
+    def get_error_payload_path(self, name='frontend.json'):
+        return super(ClientSideBaseTest, self).get_error_payload_path(name)
+
+    def upload_sourcemap(self, file_name='bundle_no_mapping.js.map',
+                         service_name='apm-agent-js',
+                         service_version='1.0.1',
+                         bundle_filepath='bundle_no_mapping.js.map',
+                         expected_ct=1):
+        path = os.path.abspath(os.path.join(self.beat_path,
+                                            'tests',
+                                            'data',
+                                            'valid',
+                                            'sourcemap',
+                                            file_name))
+        f = open(path)
+        r = requests.post(self.sourcemap_url,
+                          files={'sourcemap': f},
+                          data={'service_version': service_version,
+                                'bundle_filepath': bundle_filepath,
+                                'service_name': service_name
+                                })
+        return r
+
+    def wait_for_sourcemaps(self, expected_ct=1):
+        idx = self.smap_index_name + "*"
+        self.wait_until(
+            lambda: (self.es.count(index=idx, body={
+                "query": {"term": {"processor.name": 'sourcemap'}}}
+            )['count'] == expected_ct)
+        )
+
+    def check_error_smap(self, updated, expected_err=None, count=1):
+        rs = self.es.search(index=self.index_name, body={
+            "query": {"term": {"processor.event": "error"}}})
+        assert rs['hits']['total'] == count, "found {} documents, expected {}".format(rs['hits']['total'], count)
+        for doc in rs['hits']['hits']:
+            err = doc["_source"]["error"]
+            if "exception" in err:
+                self.check_smap(err["exception"], updated, expected_err)
+            if "log" in err:
+                self.check_smap(err["log"], updated, expected_err)
+
+    def check_transaction_smap(self, updated, expected_err=None, count=1):
+        rs = self.es.search(index=self.index_name, body={
+            "query": {"term": {"processor.event": "span"}}})
+        assert rs['hits']['total'] == count, "found {} documents, expected {}".format(rs['hits']['total'], count)
+        for doc in rs['hits']['hits']:
+            span = doc["_source"]["span"]
+            self.check_smap(span, updated, expected_err)
+
+    def check_smap(self, doc, updated, err=None):
+        if "stacktrace" not in doc:
+            return
+        for frame in doc["stacktrace"]:
+            smap = frame["sourcemap"]
+            if err is None:
+                assert 'error' not in smap
+            else:
+                assert err in smap["error"]
+            assert smap["updated"] == updated
```

Comment:
It would be good to have an assertion that the `abs_path`, `lineno` etc. is set to the values we expect them during the sourcemap process. Maybe there is a test for that, that I overlooked?

Not a blocker, in any case.

Reply:
Very good point, I think I am lacking tests in the actual sourcemap mapping, will add them.

Reply:
Actually I did test the actual mapping here https://github.com/simitt/apm-server/blob/269ee82205b3072c02a307fe3cac4643960537c8/model/stacktrace_frame_test.go#L222
by using a test sourcemap https://github.com/simitt/apm-server/blob/269ee82205b3072c02a307fe3cac4643960537c8/model/stacktrace_frame_test.go#L32